### DISCUSSION TITLE
Wait Strategy: ForAll Design

### DISCUSSION THREAD

**Discussion by hhsnopek** [testcontainers/testcontainers-go#559] [General]
## Context

This discussion reviews the existing implementation of `wait.ForAll` and is meant to answer existing questions. Any feature requests and enhancements can additionally be discussed, but should the topic should remain focused on `wait.ForAll`.

**NOTE**: For brevity the `wait.ForAll` will be described as an interface. Below is the up-to-date reference of both `Strategy` and the discussion interface describing `wait.ForAll`.

```go
type Strategy interface {
  WaitUntilReady(ctx context.Context, target StrategyTarget) (err error)
}

type ForAllStrategy interface {
  WaitUntilReady(ctx context.Context, target StrategyTarget) (err error)
  WithStartupTimeout(time.Duration) StrategyTarget
}
```

## Question: How does `ForAllStrategy#WithStartupTimeout` behave?

In the existing implementation, `ForAllStrategy#WithStartupTimeout` behaves as a deadline for all child wait strategies. This means that if we define a startup timeout of `120s` on the `ForAll` strategy all children will have their default startup timeout of `60s` or their explicit timeout.

```go
wait.ForAll(
  wait.ForLog("Ready"),                                           // Timeout: 60s (default)
  wait.ForExposedPort().WithStartupTimeout(180*time.Second),      // Timeout: 180s
  wait.ForListeningPort(port).WithStartupTimeout(10*time.Second), // Timeout: 10s
).WithStartupTimeout(120*time.Second) // If the collective startup takes longer than 120s all fail
```

## Problems

* Unless explicitly defining the timeout for all child wait strategies the `ForAll#WithStartupTimeout` acts as a Deadline
* The StartupTimeout cannot be ignored so that child wait strategies with defined startup timeouts are preferred, it defaults to `60s`.

## Comparing Wait Strategies to `testcontainers-java`

`testcontainers-java` supplies [3 modes of functionality](https://github.com/testcontainers/testcontainers-java/blob/08f75a5217f601439ef426b059908164a02deba4/core/src/main/java/org/testcontainers/containers/wait/strategy/WaitAllStrategy.java#L12-L33).

* `WITH_OUT_TIMEOUT` - This timeout is applied to each individual strategy so that the container waits for a maximum of duration
* `WITH_INDIVIDUAL_TIMEOUTS_ONLY` - This disables the outer timeout and waits for all inner strategies to hit their timeouts before exiting
* `WITH_MAXIMUM_OUTER_TIMEOUT` - The inner strategies wait with their preconfigured timeout individually and the wait all strategy kills them, if the outer limit is reached

Currently, in `testcontainers-go` there is only support for `WITH_MAXIMUM_OUTER_TIMEOUT` which is defined as the existing `ForAllStrategy#WithStartupTimeout` behavior. We are unable to have `WITH_OUT_TIMEOUT` or `WITH_INDIVIDUAL_TIMEOUTS_ONLY`.

## Suggested Design

To support all three modes I have two suggested designs. One that would require a breaking change and a second which does not.

### Design which includes a breaking change

```go
type Strategy interface {
	WaitUntilReady(context.Context, StrategyTarget) error
}

type ForAllStrategy interface {
  WithStartupTimeout(time.Duration) StrategyTarget  // Defines the default StartupTimeout for strategies which do not explcitly define a StartupTimeout
  WithDeadlineTimeout(time.Duration) StrategyTarget // Defines the deadline for all strategies
}

// Usage
wait.ForAll(
  wait.ForLog("database system is ready to accept connections").WithOccurrence(2),
  wait.ForAll(
    wait.ForListeningPort(port),
    wait.ForSQL(port, drive, URLBuilderFn).WithQuery(";"),
  ).WithDeadline(10*time.Second),
).WithStartupTimeout(20*time.Second).
  WithDeadline(120*time.Second)
```

##### Mode: WITH_INDIVIDUAL_TIMEOUTS_ONLY

```go
wait.ForAll(
  wait.ForLog("Ready"),                                           // Timeout: 60s (default)
  wait.ForExposedPort().WithStartupTimeout(180*time.Second),      // Timeout: 180s
  wait.ForListeningPort(port).WithStartupTimeout(10*time.Second), // Timeout: 10s
)
```

##### Mode: WITH_OUTER_TIMEOUT

```go
wait.ForAll(
  wait.ForLog("Ready"),                                           // Timeout: 120s (from ForAll.WithStartupTimeout)
  wait.ForExposedPort().WithStartupTimeout(180*time.Second),      // Timeout: 180s
  wait.ForListeningPort(port).WithStartupTimeout(10*time.Second), // Timeout: 10s
).WithStartupTimeout(120*time.Second)
```

##### Mode: WITH_MAXIMUM_OUTER_TIMEOUT
```go
wait.ForAll(
  wait.ForLog("Ready"),                                           // Timeout: 60s (default)
  wait.ForExposedPort().WithStartupTimeout(180*time.Second),      // Timeout: 180s
  wait.ForListeningPort(port).WithStartupTimeout(10*time.Second), // Timeout: 10s
).WithDeadline(360*time.Second)
```

##### All modes together
```go
wait.ForAll(
  wait.ForLog("Ready"),                                           // Timeout: 120s (from ForAll.WithStartupTimeout)
  wait.ForExposedPort().WithStartupTimeout(180*time.Second),      // Timeout: 180s
  wait.ForListeningPort(port).WithStartupTimeout(10*time.Second), // Timeout: 10s
).WithStartupTimeout(120*time.Second).                            // Applies default StartupTimeout when not explictly defined
  WithDeadline(360*time.Second)                                   // Applies deadline for all Wait Strategies
```

### Design with backwards compatibility
```go
type ForAllStrategy interface {
  WithStartupTimeoutDefault(time.Duration) StrategyTarget  // Defines the default StartupTimeout for strategies which do not explcitly define a StartupTimeout
  WithDeadlineTimeout(time.Duration) StrategyTarget        // Defines the deadline for all strategies
  WithStartupTimeout(time.Duration) StrategyTarget         // Deprecated: use ForAllStrategy.UseDeadlineTimeout instead.
}

// usage
wait.ForAll(
  wait.ForLog("database system is ready to accept connections").WithOccurrence(2),
  wait.ForAll(
    wait.ForListeningPort(port),
    wait.ForSQL(port, drive, URLBuilderFn).WithQuery(";"),
  ).WithDeadline(10*time.Second),
).WithStartupTimeoutDefault(20*time.Second).
  WithDeadline(120*time.Second)
```

**Comment by mdelapenya** [comment]
Thanks for such a detailed design issue. We highly appreciate your time here.

I'd say that the breaking-change alternative seems cleaner to me, even though it comes with the BC. Said that, providing nice docs and good examples about the new usage, would be a perfect combo for the BC alternative.

**Comment by hhsnopek** [comment]
I took a look at #322 and took notice to the rename of `WithStartupTimeout` to `WithTimeout`. If we were to utilize this language instead of the purposed language it may be more clear with the following interface:

```go
type ForAllStrategy interface {
  WithTimeoutDefault(time.Duration) StrategyTarget  // Defines the default StartupTimeout for strategies which do not explcitly define a StartupTimeout
  WithDeadline(time.Duration) StrategyTarget        // Defines the deadline for all strategies
}

// usage
wait.ForAll(
  wait.ForLog("database system is ready to accept connections").WithOccurrence(2),
  wait.ForAll(
    wait.ForListeningPort(port),
    wait.ForSQL(port, drive, URLBuilderFn).WithQuery(";"),
  ).WithDeadline(10*time.Second),
).WithTimeoutDefault(20*time.Second).
  WithDeadline(120*time.Second)
```

Thoughts?

**Reply by mdelapenya** [reply]
Indeed, thanks for connecting with that issue.

I think we have a consistent design although I'd like to check my mates @eddumelendez @HofmeisterAn @kiview first. wdyt?

**Comment by mdelapenya** [comment]
This discussion can be considered solved, as it has been merged and released into v0.16.0
